trin-core/src/portalnet/overlay_service.rs

```diff
@@ -365,10 +389,64 @@ impl OverlayService {
                     }
                 }
                 _ = OverlayService::bucket_maintenance_poll(self.protocol.clone(), &self.kbuckets) => {}
+                _ = bucket_refresh_interval.tick() => {
+                    debug!("Bucket refresh lookup");
+                    self.bucket_refresh_lookup();
+                }
             }
         }
     }
 
+    /// Generate random NodeId based on bucket index target.
+    /// First we generate a random distance metric with leading zeroes based on the target bucket.
+    /// Then we XOR the result distance with the local NodeId to get the random target NodeId
+    fn generate_random_node_id(&self, target_bucket_idx: u8) -> Result<NodeId, RandomNodeIdError> {
+        let distance_leading_zeroes = 255 - target_bucket_idx;
+        let random_distance = binary_string::generate_random(distance_leading_zeroes)?;
+        let random_distance = binary_string::to_byte_array(&random_distance)?;
```

Comment:
Is there a reason for the intermediate generation of a string? It seems that we should be able to generate a random byte array directly, and this function should never fail.

Reply:
We are able to generate a random byte array directly but we want to have control over the leading zeroes in the 256-bit value. This means we will need additional logic to decide what byte value range we want for every 8 bits. For example, if we want to generate a random 32 size byte array with N leading bit zeroes, we need to do something like:

```python
first_zero_bytes = N // 8
leading_zeroes_in_next_byte = N % 8

If leading_zeroes_in_next_byte == 7
   next_byte_value = 1
elif leading_zeroes_in_next_byte == 6
   next_byte_value = random.randrange(3)
elif leading_zeroes_in_next_byte == 5
   next_byte_value = random.randrange(5)
elif leading_zeroes_in_next_byte == 4
   next_byte_value = random.randrange(9)
 .......
 ```
 
 Then we have to prefix the final result with the first zero bytes, append the next non-zero byte, and finally generate random bytes for the rest of the array.
 
Not sure if this or the current implementation using a binary string is more clear/performant. Happy to change it if we agree on it.

Reply:
I'm less concerned with performance and more concerned with unnecessary error handling.

I was thinking we could something like the following.

```rust
fn random_32_byte_array(leading_zeros: usize) -> Vec<u8> {
    let first_zero_bytes = leading_zeros / 8;
    let first_nonzero_byte_leading_zeros = leading_zeros % 8;

    let mut bytes: Vec<u8> = vec![];
    for i in 0..first_zero_bytes {
        vec.push(0);
    }

    let mut random_bytes = [0u8, 32 - first_zero_bytes];
    rand::thread_rng().fill_bytes(&mut random_bytes);
    random_bytes[0] &= (u8::MAX >> first_nonzero_byte_leading_zeros);

    bytes.append(random_bytes.into());

    bytes
}
```

Not sure that the code above works as written, but essentially use some bitwise operations to do this.

Reply:
It turns out the logic should be a little more complex but I updated the PR to generate a random 32-byte array without intermediate binary string generation. Let me know what you think!